x/evidence/legacy/v0_40/migrate_test.go

```diff
@@ -0,0 +1,46 @@
+package v040_test
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/require"
+
+	"github.com/cosmos/cosmos-sdk/codec"
+	cryptocodec "github.com/cosmos/cosmos-sdk/crypto/codec"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	v038evidence "github.com/cosmos/cosmos-sdk/x/evidence/legacy/v0_38"
+	v040evidence "github.com/cosmos/cosmos-sdk/x/evidence/legacy/v0_40"
+	"github.com/cosmos/cosmos-sdk/x/evidence/types"
+)
+
+func TestMigrate(t *testing.T) {
+	v040Codec := codec.New()
+	cryptocodec.RegisterCrypto(v040Codec)
+
+	addr1, _ := sdk.AccAddressFromBech32("cosmos1xxkueklal9vejv9unqu80w9vptyepfa95pd53u")
+
+	evidenceGenState := v038evidence.GenesisState{
+		Params: v038evidence.Params{MaxEvidenceAge: v038evidence.DefaultMaxEvidenceAge},
+		Evidence: []v038evidence.Evidence{&types.Equivocation{
+			Height:           20,
+			Power:            100,
+			ConsensusAddress: addr1.Bytes(),
+		}},
+	}
+
+	migrated := v040evidence.Migrate(evidenceGenState)
+	expected := `{
+  "evidence": [
+    {
+      "height": "20",
+      "time": "0001-01-01T00:00:00Z",
+      "power": "100",
+      "consensus_address": "cosmosvalcons1xxkueklal9vejv9unqu80w9vptyepfa99x2a3w"
+    }
+  ]
+}`
+
+	bz, err := v040Codec.MarshalJSONIndent(migrated, "", "  ")
```

Comment:
I have an error on this line:

```
JSON marshal marshaling error for {TypeUrl:/cosmos.evidence.Equivocation Value:[8 20 18 11 8 128 146 184 195 152 254 255 255 255 1 24 100 34 20 49 173 204 219 253 249 89 153 48 188 152 56 119 184 172 10 201 144 167 165] XXX_NoUnkeyedLiteral:{} XXX_unrecognized:[] XXX_sizecache:0 cachedValue:0xc0000aa480 compat:<nil>}, this is likely because amino is being used directly (instead of codec.Codec which is preferred) or UnpackInterfacesMessage is not defined for some type which contains a protobuf Any either directly or via one of its members. To see a stacktrace of where the error is coming from, set the var Debug = true in codec/types/compat.go
```

Not sure what to do here, should I use `v040Codec := codec.NewProtoCodec()` and register the interface/impls for Evidence & Equivocation?

Reply:
for 0.40, is genesis state expected to be amino or proto encoded?

Reply:
My general thought is that it should follow the chain configuration either amino or proto. We could have a flag to specify which format, but the simplest would be to just use the JSONMarshaler and TxConfig on `client.Context`.

Reply:
> but the simplest would be to just use the JSONMarshaler and TxConfig on client.Context.

I would opt for this for 0.40 migration logic. The fields can be set on the `migrate` command flags.

Reply:
The challenge now is with interface type registration is that right @amaurymartiny ?

Reply:
Interface registration is done in https://github.com/cosmos/cosmos-sdk/pull/6839/commits/766bb7ea3a2cc4e1cefd6cb7c1cdbd229ef35f23.

If the flag on `migrate` is optional, we could maybe add it in a follow-up PR?